test/helpers/kind/setup.go

```diff
@@ -19,91 +19,119 @@ package kind
 import (
 	"bufio"
 	"bytes"
-	"flag"
 	"fmt"
 	"io"
-	"io/ioutil"
 	"os"
 	"os/exec"
-	"strings"
 	"sync"
 
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
-
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/gomega"
 	"k8s.io/client-go/kubernetes"
 	restclient "k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/cluster-api/util"
 )
 
-const (
-	kindestImage = "kindest/node:v1.16.4"
-)
-
-var (
-	kindBinary    = flag.String("kindBinary", "kind", "path to the kind binary")
-	kubectlBinary = flag.String("kubectlBinary", "kubectl", "path to the kubectl binary")
-)
+const defaultImage = "kindest/node:v1.17.2"
 
-// Cluster represents the running state of a KIND cluster.
+// Cluster represents the running state of a Kind cluster.
 // An empty struct is enough to call Setup() on.
 type Cluster struct {
-	Name     string
-	tmpDir   string
-	kubepath string
+	Name           string
+	KubeconfigPath string
+	Image          string
+	Retain         bool
+	Wait           string
+	kindBinary     string
+	kubectlBinary  string
 }
 
-// Setup creates a kind cluster and returns a path to the kubeconfig
-// nolint:gosec
+// Setup creates a Kind cluster
 func (c *Cluster) Setup() {
+	c.init()
+	fmt.Fprintf(ginkgo.GinkgoWriter, "creating Kind cluster named %q\n", c.Name)
+
+	cmd := exec.Command(
+		c.kindBinary,
+		"create", "cluster",
+		"--name", c.Name,
+		"--image", c.Image,
+		"--kubeconfig", c.KubeconfigPath,
+		"--wait", c.Wait,
+	)
+	if c.Retain {
+		cmd.Args = append(cmd.Args, "--retain")
+	}
+	c.run(cmd)
+}
+
+func (c *Cluster) init() {
+	if c.Name == "" {
+		c.Name = "kind-test-" + util.RandomString(6)
+	}
+	if c.KubeconfigPath == "" {
+		c.KubeconfigPath = fmt.Sprintf("kubeconfig-%s", c.Name)
+	}
+	if c.Image == "" {
+		c.Image = defaultImage
+	}
+	if c.Wait == "" {
+		c.Wait = "0"
+	}
 	var err error
-	c.tmpDir, err = ioutil.TempDir("", "kind-home")
-	ExpectWithOffset(1, err).NotTo(HaveOccurred())
-	fmt.Fprintf(GinkgoWriter, "creating Kind cluster named %q\n", c.Name)
-	c.run(exec.Command(*kindBinary, "create", "cluster", "--image", kindestImage, "--name", c.Name))
-	path := c.runWithOutput(exec.Command(*kindBinary, "get", "kubeconfig-path", "--name", c.Name))
-	c.kubepath = strings.TrimSpace(string(path))
-	fmt.Fprintf(GinkgoWriter, "kubeconfig path: %q. Can use the following to access the cluster:\n", c.kubepath)
-	fmt.Fprintf(GinkgoWriter, "export KUBECONFIG=%s\n", c.kubepath)
+	c.kindBinary, err = exec.LookPath("kind")
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
```

Comment:
Does it makes sense to use ginko inside helpers/framework in light of https://github.com/kubernetes-sigs/cluster-api/issues/2306?
@detiber @chuckha ^^

Reply:
These helpers will eventually be going away in favor of the e2e test framework.